File: rs/src/bin/124_binary_tree_maximum_path_sum.rs
use std::cell::RefCell;
use std::rc::Rc;

use tml::TreeNode;

fn main() {
    let root = TreeNode {
        val: -10,
        left: Some(Rc::new(RefCell::new(TreeNode::new(9)))),
        right: Some(Rc::new(RefCell::new(TreeNode {
            val: 20,
            left: Some(Rc::new(RefCell::new(TreeNode::new(15)))),
            right: Some(Rc::new(RefCell::new(TreeNode::new(7)))),
        }))),
    };

    println!("{:?}", max_path_sum(Some(Rc::new(RefCell::new(root)))));
}

fn max_path_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    let mut num = i32::MIN;

    fn max_gain(max_sum: &mut i32, root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        match root {
            Some(rt) => {
                // 计算左边分支最大值，如果为负则不选择
                let left_gain = 0.max(max_gain(max_sum, rt.borrow().left.clone()));
                // 同理计算右边分支
                let right_gain = 0.max(max_gain(max_sum, rt.borrow().right.clone()));
                // 比较现有的最大值与计算后的最大值之和，将其赋值给 max_sum
                *max_sum = (*max_sum).max(rt.borrow().val + left_gain + right_gain);
                // 返回当前节点的值与左右分支中最大值之和
                rt.borrow().val + left_gain.max(right_gain)
            }
            None => 0,
        }
    }

    max_gain(&mut num, root);
    num
}

File: go/problems/124_binary_tree_maximum_path_sum.go
package problems

import (
	"math"
	"tml/util"
)

func MaxPathSum_124(root *util.TreeNode) int {
	var num = math.MinInt32
	max_gain(&num, root)
	return num
}

func max_gain(num *int, node *util.TreeNode) int {
	if node == nil {
		return 0
	}

	left := max_gain(num, node.Left)
	right := max_gain(num, node.Right)

	if left < 0 {
		left = 0
	}

	if right < 0 {
		right = 0
	}

	*num = max(*num, left+right+node.Val)

	return max(left, right) + node.Val
}
